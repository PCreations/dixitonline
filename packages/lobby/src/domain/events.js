export const newGameCreatedEvent = ({ gameId }) => ({
  type: '[lobby] - a new game has been created',
  payload: {
    gameId,
  },
});

export const playerJoinedGame = ({ gameId, playerId }) => ({
  type: '[lobby] - a new player has joined a game',
  payload: {
    gameId,
    playerId,
  },
});
