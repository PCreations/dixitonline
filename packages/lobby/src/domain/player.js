export const makePlayer = ({ id, name } = {}) => {
  if (!id) throw new Error('Player must contain an id');
  if (!name) throw new Error('Player must have a name');
  return { id, name };
};

export const equals = (player1, player2) => player1.id === player2.id;
