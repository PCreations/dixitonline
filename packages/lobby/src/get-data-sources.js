import { makeLobbyRepository } from './lobby-repository';

export const makeGetDataSources = ({ lobbyRepository = makeLobbyRepository() } = {}) => () => ({
  lobbyRepository,
});
