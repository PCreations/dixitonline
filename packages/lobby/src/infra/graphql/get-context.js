export const makeGetContext = ({ dispatchDomainEvents, authorizationService } = {}) => async ({ req }) => ({
  dispatchDomainEvents,
  currentUser: await authorizationService.getCurrentUser(req),
});
