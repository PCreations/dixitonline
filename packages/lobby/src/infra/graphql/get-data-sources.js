import { makeLobbyRepository } from '../../repos/lobby-repository';

export const makeGetDataSources = ({ lobbyRepository = makeLobbyRepository() } = {}) => () => ({
  lobbyRepository,
});
