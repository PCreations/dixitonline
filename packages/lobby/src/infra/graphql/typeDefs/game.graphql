type LobbyGame {
  id: ID!
  host: LobbyPlayer!
}
