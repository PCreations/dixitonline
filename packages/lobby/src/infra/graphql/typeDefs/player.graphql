type LobbyPlayer {
  id: ID!
  name: String!
}
