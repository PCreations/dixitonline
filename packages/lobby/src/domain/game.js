/* eslint-disable max-classes-per-file */
import { playerJoinedGame } from './events';
import { equals as playerEquals } from './player';

export class GameAlreadyJoinedByPlayerError extends Error {}
export class MaximumNumberOfPlayerReachedError extends Error {}

export const MAXIMUM_NUMBER_OF_PLAYERS = 6;

export const makeGame = ({ id, host, players = [] } = {}) => {
  if (!id) throw new Error('Game must contain an id');
  if (!host) throw new Error('Game must have an host');

  return Object.freeze({
    id,
    host,
    players,
  });
};

const getAllPlayers = game => [game.host, ...game.players];

const isGameFull = game => getAllPlayers(game).length === MAXIMUM_NUMBER_OF_PLAYERS;

export const joinPlayer = (game, player) => {
  if (getAllPlayers(game).some(playerEquals.bind(null, player))) throw new GameAlreadyJoinedByPlayerError();
  if (isGameFull(game)) throw new MaximumNumberOfPlayerReachedError();
  return [
    makeGame({ ...game, players: game.players.concat(player) }),
    [playerJoinedGame({ gameId: game.id, playerId: player.id })],
  ];
};
