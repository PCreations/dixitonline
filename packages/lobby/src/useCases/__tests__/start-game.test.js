import gql from 'graphql-tag';
import { createTestClient } from 'apollo-server-testing';
import { makeTestServer } from '../../__tests__/test-server';
import { makeGetDataSources } from '../../infra/graphql/get-data-sources';
import { buildTestGame } from '../../__tests__/dataBuilders/game';
import { buildLobbyRepositoryInitialGames } from '../../__tests__/dataBuilders/lobby-repository-initial-games';
import { makeNullLobbyRepository, GameNotFoundError } from '../../repos';
import { buildTestPlayer } from '../../__tests__/dataBuilders/player';
import { newGameStartedEvent } from '../../domain/events';

describe('start game', () => {
  test('a game with more than 2 players can be started by the host', async () => {
    // arrange
    const game = buildTestGame()
      .withId('g42')
      .withXPlayers(3)
      .build();
    const initialGames = buildLobbyRepositoryInitialGames()
      .withGames([game])
      .build();
    const lobbyRepository = makeNullLobbyRepository({
      gamesData: initialGames,
    });
    const dispatchDomainEvents = jest.fn();
    const server = makeTestServer({
      getDataSources: makeGetDataSources({
        lobbyRepository,
      }),
      dispatchDomainEvents,
      currentUserId: game.host.id,
      currentUserUsername: game.host.name,
    });
    const LOBBY_START_GAME = gql`
      mutation LobbyStartGame($lobbyStartGameInput: LobbyStartGameInput!) {
        lobbyStartGame(lobbyStartGameInput: $lobbyStartGameInput) {
          ... on LobbyStartGameResultSuccess {
            gameId
          }
        }
      }
    `;

    // act
    const { mutate } = createTestClient(server);
    const response = await mutate({
      mutation: LOBBY_START_GAME,
      variables: { lobbyStartGameInput: { gameId: 'g42' } },
      operationName: 'LobbyStartGame',
    });

    // assert
    expect.assertions(3);
    expect(response).toMatchSnapshot();
    await expect(lobbyRepository.getGameById('g42')).rejects.toEqual(new GameNotFoundError('g42'));
    expect(dispatchDomainEvents).toHaveBeenCalledWith([
      newGameStartedEvent({ gameId: 'g42', playerIds: [game.host.id, ...game.players.map(({ id }) => id)] }),
    ]);
  });
  test("a player can't start a game if she's not the host", async () => {
    // arrange
    const game = buildTestGame()
      .withId('g42')
      .withXPlayers(3)
      .build();
    const randomPlayer = buildTestPlayer().build();
    const initialGames = buildLobbyRepositoryInitialGames()
      .withGames([game])
      .build();
    const lobbyRepository = makeNullLobbyRepository({
      gamesData: initialGames,
    });
    const dispatchDomainEvents = jest.fn();
    const server = makeTestServer({
      getDataSources: makeGetDataSources({
        lobbyRepository,
      }),
      dispatchDomainEvents,
      currentUserId: randomPlayer.id,
      currentUserUsername: randomPlayer.name,
    });
    const LOBBY_START_GAME = gql`
      mutation LobbyStartGame($lobbyStartGameInput: LobbyStartGameInput!) {
        lobbyStartGame(lobbyStartGameInput: $lobbyStartGameInput) {
          ... on LobbyStartGameResultError {
            type
          }
        }
      }
    `;

    // act
    const { mutate } = createTestClient(server);
    const response = await mutate({
      mutation: LOBBY_START_GAME,
      variables: { lobbyStartGameInput: { gameId: 'g42' } },
      operationName: 'LobbyStartGame',
    });

    // assert
    expect.assertions(2);
    expect(response).toMatchSnapshot();
    await expect(lobbyRepository.getGameById('g42')).resolves.toEqual(game);
  });
});
