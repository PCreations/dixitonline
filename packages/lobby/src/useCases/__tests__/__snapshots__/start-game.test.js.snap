// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`start game a game with more than 2 players can be started by the host 1`] = `
Object {
  "data": Object {
    "lobbyStartGame": Object {
      "gameId": "g42",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
